src/zmqsubscriber/zmqsubscriber.go

```diff
@@ -155,6 +119,100 @@ func (z *ZMQSubscriber) Run() error {
 	return nil
 }
 
+func (z *ZMQSubscriber) processMessage(topic string, payload [][]byte) error {
+	// TODO: use GetTime() and allow other time sources (eg NTP-corrected)
+	firstSeen := time.Now().UTC()
+
+	switch topic {
+	case topicHashTx:
+		tx, err := parseTransactionHash(firstSeen, payload)
+		if err != nil {
+			return err
+		}
+		z.IncomingTx <- *tx
+	case topicRawTxWithFee:
+		tx, err := parseTransaction(firstSeen, payload)
+		if err != nil {
+			return err
+		}
+		z.IncomingTx <- *tx
+	case topicRawBlock:
+		block, err := parseBlock(firstSeen, payload)
+		if err != nil {
+			return err
+		}
+		z.IncomingBlocks <- *block
+	default:
+		return fmt.Errorf("unknown topic %s", topic)
+	}
+
+	return nil
+}
+
+// Stop sets the cancel flag to true. The ZMQSubscriber is stopped after it finishes
+// receiving a message or
 func (z *ZMQSubscriber) Stop() {
 	z.cancel = true
 }
+
+func parseTransactionHash(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
+	}
+
+	// payload[1] contains a 16bit LE sequence number provided by Bitcoin Core,
+	// which is not used here, but noted for completeness.
+	txhash, _ := payload[0], payload[1]
+
+	var txid types.Hash32
+	copy(txid[:], txhash)
+
+	return &types.Transaction{
+		FirstSeen: firstSeen,
+		TxID:      txid,
+	}, nil
+}
+
+func parseTransaction(firstSeen time.Time, payload [][]byte) (*types.Transaction, error) {
+	if len(payload) != 2 {
+		return nil, fmt.Errorf("unexpected payload length: expected len((tx hash, sequence) == 2 but got len(payload) == %d", len(payload))
+	}
+
+	// payload[1] contains a 16bit LE sequence number provided by Bitcoin Core,
+	// which is not used here, but noted for completeness.
+	// rawtxwithfee is the rawtx by Bitcoin Core concatinated with the 8 byte BE
+	// transaction fee
+	rawtxwithfee, _ := payload[0], payload[1]
+
+	length := len(rawtxwithfee)
+	if length <= 8 {
+		return nil, errors.New("unexpected rawtxwithfee length")
+	}
+	rawtx, feeBytes := rawtxwithfee[:length-8], rawtxwithfee[length-8:]
+
+	wireTx := wire.NewMsgTx(wire.TxVersion)
+	if err := wireTx.Deserialize(bytes.NewReader(rawtx)); err != nil {
+		return nil, fmt.Errorf("could not deserialize the rawtx as wire.MsgTx: %s", err)
+	}
+
+	var txid types.Hash32
+	wireTxHash := wireTx.TxHash()
+	copy(txid[:], []byte(wireTxHash.CloneBytes()))
+
+	fee := binary.BigEndian.Uint64(feeBytes)
+
+	vsizeFloat := (float64(wireTx.SerializeSizeStripped()*3) + float64(wireTx.SerializeSize())) / 4
+	vsize := int(math.Ceil(vsizeFloat))
```

Comment:
how about we track the weight units instead and keep full precision?

Reply:
Good point!

Reply:
Changed in https://github.com/0xB10C/bademeister-go/pull/19/commits/08d2148b10a7831418f37b30d78a5e1ae872d65d